safehttp/plugins/fetch_metadata/fetch_metadata_test.go

```diff
@@ -0,0 +1,411 @@
+			wantBody:    "",
+			wantReport:  "",
+		},
+		{
+			name: "Sec-Fetch-Site: cross-site",
+			req: func() *safehttp.IncomingRequest {
+				req := safehttptest.NewRequest("POST", "/pizza", nil)
+				req.Header.Add("Sec-Fetch-Site", "cross-site")
+				return req
+			}(),
+			wantStatus:  200,
+			wantHeaders: map[string][]string{},
+			wantBody:    "",
+			wantReport:  "POST /pizza",
+		},
+		{
+			name: "POST request with Sec-Fetch-Mode: navigate from image",
+			req: func() *safehttp.IncomingRequest {
+				req := safehttptest.NewRequest("POST", "/pizza", nil)
+				req.Header.Add("Sec-Fetch-Site", "cross-site")
+				req.Header.Add("Sec-Fetch-Mode", "navigate")
+				req.Header.Add("Sec-Fetch-Dest", "image")
+				return req
+			}(),
+			wantStatus:  200,
+			wantHeaders: map[string][]string{},
+			wantBody:    "",
+			wantReport:  "POST /pizza",
+		},
+		{
+			name: "GET request with Sec-Fetch-Mode: navigate from object",
+			req: func() *safehttp.IncomingRequest {
+				req := safehttptest.NewRequest("GET", "/pizza", nil)
+				req.Header.Add("Sec-Fetch-Site", "cross-site")
+				req.Header.Add("Sec-Fetch-Mode", "navigate")
+				req.Header.Add("Sec-Fetch-Dest", "object")
+				return req
+			}(),
+			wantStatus:  200,
+			wantHeaders: map[string][]string{},
+			wantBody:    "",
+			wantReport:  "GET /pizza",
+		},
+		{
+			name: "GET request with Sec-Fetch-Mode: navigate from embed",
+			req: func() *safehttp.IncomingRequest {
+				req := safehttptest.NewRequest("GET", "/pizza", nil)
+				req.Header.Add("Sec-Fetch-Site", "cross-site")
+				req.Header.Add("Sec-Fetch-Mode", "navigate")
+				req.Header.Add("Sec-Fetch-Dest", "embed")
+				return req
+			}(),
+			wantStatus:  200,
+			wantHeaders: map[string][]string{},
+			wantBody:    "",
+			wantReport:  "GET /pizza",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			logger := &fooLog{}
+			p := fetchmetadata.NewPlugin()
+			p.SetReportMode(logger)
+			rec := safehttptest.NewResponseRecorder()
+
+			p.Before(rec.ResponseWriter, test.req)
+
+			if status := rec.Status(); status != test.wantStatus {
+				t.Errorf("status code got: %v want: %v", status, test.wantStatus)
+			}
+			if diff := cmp.Diff(test.wantHeaders, map[string][]string(rec.Header())); diff != "" {
+				t.Errorf("rec.Header() mismatch (-want +got):\n%s", diff)
+			}
+			if got := rec.Body(); got != test.wantBody {
+				t.Errorf("response body got: %q want: %q", got, test.wantBody)
+			}
+			if logger.report != test.wantReport {
+				t.Errorf("logger.report: got %s, want %s", logger.report, test.wantReport)
+			}
+		})
+	}
+}
+
+func TestReportModeMissingLogger(t *testing.T) {
+	p := fetchmetadata.NewPlugin()
+	defer func() {
+		if r := recover(); r == nil {
+			t.Error("p.SetReportMode(nil) expected panic")
+		}
+	}()
+	p.SetReportMode(nil)
+}
+
+type fooLog struct {
+	report string
+}
+
+func (l *fooLog) Log(r *safehttp.IncomingRequest) {
+	l.report = r.Method() + " " + r.URL.Path()
+}
```

Comment:
Move this above `TestReportModeWithLogger` since that is where it is first used.